doubly_linked_list/doubly_linked_list.py

```diff
@@ -44,25 +44,98 @@ def __len__(self):
     return self.length
 
   def add_to_head(self, value):
-    pass
+    if self.head != None:
+        self.head.insert_before(value)
+        self.head = self.head.prev
+    else:
+        self.head = ListNode(value)
+        self.tail = self.head
+    self.length += 1
+    # if not self.head and not self.tail:
+    #     new_node = ListNode(value)
+    #     self.head = new_node
+    #     self.tail = new_node
+    # else:
+    #     new_node = ListNode(value)
+    #     self.head.insert_before(new_node)
+    #     self.head = new_node
+
 
   def remove_from_head(self):
-    pass
+    if self.head == None:
+        self.tail = None
+        self.length = 0
+        head_capture = None
+    else:
+        new_head = self.head.next 
+        head_capture = self.head.value
+        self.head.delete()
+        self.head = new_head
+        self.length -= 1
+        if self.length == 0:
+            self.tail = None
+        return head_capture
 
   def add_to_tail(self, value):
-    pass
+    if self.tail != None:
+        self.tail.insert_after(value)
+        self.tail = self.tail.next
+    else:
+        self.tail = ListNode(value)
+        self.head = self.tail
+    self.length += 1
 
   def remove_from_tail(self):
-    pass
+    if self.tail == None:
+        self.head = None
+        self.length = 0
+        head_capture = None
+    else:
+        new_tail = self.tail.prev
+        head_capture = self.tail.value
+        self.tail.delete()
```

Comment:
Good job on understanding the problem! A lot of people struggled at using node.delete()